File: 39-Combination-Sum/solution.cpp
class Solution {
public:
	vector<vector<int>> ret;
	vector<int> ans;
	void helper(int dep, vector<int>& candidates, int target) {
		if (target == 0)
			ret.push_back(ans);
		if (dep == candidates.size())
			return;
		if (target < candidates[dep])
			return;
		helper(dep + 1, candidates, target);
		for (int i = 1; i <= target / candidates[dep]; ++i) {
			ans.push_back(candidates[dep]);
			helper(dep + 1, candidates, target - i * candidates[dep]);
		}
		while (ans.back() == candidates[dep])
			ans.pop_back();
	}
	vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
		ret.clear();
		ans.clear();
		sort(candidates.begin(), candidates.end());
		helper(0, candidates, target);
		return ret;
	}
};

File: 39-Combination-Sum/solution.py
class Solution(object):
    def helper(self, candidates, target, current_list, current_index, result):
        """
        :param candidates:
        :param target:
        :param current_list:
        :param current_index:
        :param result:
        :return:
        """
        if target <= 0:
            if target == 0:
                result.append(current_list)
            return -1
        else:
            for i in xrange(current_index, len(candidates)):
                v = candidates[i]
                if self.helper(candidates, target - v, current_list + [v], i, result) == -1:
                    break
        return 0

    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()
        res = []
        self.helper(candidates, target, [], 0, res)
        return res